lib/streamlit/runtime/runtime.py

```diff
@@ -190,10 +191,18 @@ def __init__(self, config: RuntimeConfig):
         self._uploaded_file_mgr.on_files_updated.connect(self._on_files_updated)
         self._media_file_mgr = MediaFileManager(storage=config.media_file_storage)
         self._cache_storage_manager = config.cache_storage_manager
+        self._script_cache = ScriptCache()
+
+        # Clear the ScriptCache when a user script file changes
+        self._sources_watcher = LocalSourcesWatcher(self._main_script_path)
+        self._sources_watcher.register_file_change_callback(
+            lambda _: self._script_cache.clear()
+        )
```

Comment:
Would it be possible to have a more targeted cache expiration function rather than one that clears the cache entirely?

In the case where a multipage app has a large number of pages, a change to one page will potentially clear out a large number of cached pages that haven't changed.

This probably isn't the biggest deal, though, given that it's not super important for everything to be as fast as possible while an app is still being developed.

Reply:
Definitely possible, but didn't seem worth the extra complexity + effort given the circumstances when it would be valuable